test/jdk/javax/crypto/KeyGenerator/CompareKeys.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import javax.crypto.SecretKey;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.security.Key;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.util.Arrays;
+import javax.crypto.KeyGenerator;
+
+/*
+ * @test
+ * @bug 8185127
+ * @summary Test key comparison between the Keys generated through KeyGenerator
+ * @run main CompareKeys
+ */
+public class CompareKeys {
+
+    public static void main(String[] args) throws Exception {
+
+        for (KeyAlgo ka : KeyAlgo.values()) {
+            SecretKey k = ka.genSecretKey();
+            checkKeyEquality(k, copy(k));
+        }
+
+        for (KeyPairAlgo kpa : KeyPairAlgo.values()) {
+            KeyPair kp = kpa.genKeyPair();
+            checkKeyPairEquality(kp, copy(kp));
+        }
+        System.out.println("Done!");
+    }
+
+    private static void checkKeyPairEquality(KeyPair origkp, KeyPair copykp)
+            throws Exception {
+
+        checkKeyEquality(origkp.getPrivate(), copykp.getPrivate());
+        checkKeyEquality(origkp.getPublic(), copykp.getPublic());
+    }
+
+    /**
+     * Compare original Key with another copy.
+     */
+    private static void checkKeyEquality(Key origKey, Key copyKey) {
+
+        if (!origKey.equals(copyKey)
+                && !Arrays.equals(origKey.getEncoded(), copyKey.getEncoded())
+                && origKey.hashCode() != copyKey.hashCode()) {
+            throw new RuntimeException("Key inequality found");
```

Comment:
Check the format equality as well?

Reply:
Should
 be || instead of &&?

Reply:
Considering equality and hashCode may not be same for all KeyTypes, i have relaxed the verification with || operator. Also added print statement for possible failure cause.

Reply:
I am not too sure about this check, please see my other comment.